src/rust/plugin-work-queue/README.md

```diff
@@ -0,0 +1,106 @@
+   tenant_id        uuid                       NOT NULL,
+   plugin_id        uuid                       NOT NULL,
+   pipeline_message bytea                      NOT NULL,
+   status           plugin_work_queue.status_t NOT NULL,
+   creation_time    timestamptz                NOT NULL,
+   last_updated     timestamptz                NOT NULL,
+   visible_after    timestamptz,
+   try_count        integer                    NOT NULL,
+   CHECK (length(pipeline_message) < plugin_work_queue.megabytes(256)),
+   CHECK (last_updated >= creation_time)
+)
+   PARTITION BY RANGE (creation_time);
+
+CREATE INDEX IF NOT EXISTS execution_key_ix ON plugin_work_queue.generator_plugin_executions (execution_key);
+CREATE INDEX IF NOT EXISTS creation_time_ix ON plugin_work_queue.generator_plugin_executions (creation_time);
+
+CREATE INDEX IF NOT EXISTS execution_key_ix ON plugin_work_queue.analyzer_plugin_executions (execution_key);
+CREATE INDEX IF NOT EXISTS creation_time_ix ON plugin_work_queue.analyzer_plugin_executions (creation_time);
+```
+
+`execution_key` is a `bigserial` - an auto-incrementing 64bit integer.
+
+`plugin_id` is the uuid name of the plugin to send this execution result to.
+
+`pipeline_message` is the raw bytes to be interpreted by the plugin. The message is limited to 256MB as an arbitrary but
+reasonable upper limit.
+
+`status` is an enum representing the state. 'enqueued' is the default, and a message will be set to either 'processed' or 'failed' based on its success.
+
+`creation_time` is when the row was created.
+
+`last_updated` is set with each update to the row
+
+`visible_after` is essentially a visibility timeout. A `visible_after` of NULL means a message is available, which
+is the default state. Every time a message is SELECT'd we update `visible_after`. See the `Visibility Timeout` section
+below.
+
+`try_count` on every receive we increment `try_count` to indicate another attempt to process this message
+
+
+## Algorithm
+The query for processing a message is:
+
+1. Find the oldest message that
+    1. Has not aged out (1 day)
+    2. Is not currently being locked by another request
+    3. Is 'enqueued'
+    4. Is visible ie: the `visible_after` is null or < CURRENT_TIMESTAMP
+2. Update that message
+    1. Increment `try_count`
+    2. Update `visible_after` to CURRENT_TIMESTAMP + an interval
+3. Return the message
+
+The message is then processed by the consumer.
+
+If the message is successfully processed or if it fails, update the row:
+1. Set the `execution_result`
+2. Set the `last_updated` to CURRENT_TIMESTAMP
+3. Set the `status` to `processed` or `failed` accordingly
+
+Otherwise, if the message is not successfully processed but can be retried,
+do nothing. It will be picked up again later.
+
+Note that just because a message has the `enqueued` status does *not* mean that it will be processed. We don't
+guarantee that, on failure, we'll mark the `status` as failed.
```

Comment:
Why? This seems important to spell out explicitly.

Also, this sounds like two separate things. "Not being processed" sounds like we can have a message that is submitted to the queue, but is never once picked up by a worker.

Never marking something as failed seems like the second thing. We may want to think about another failure state for this. The current `failed` status arises when we receive an acknowledgement from the worker that, indeed, it tried to process this message, but it did not work. The "new" failure status seems like it would involve exceeding a retry count. That is, we keep sending this message out to workers, but we never get a response back. As a result, we continue to increment the try count. After some maximum is exceeded, we probably need to mark this message as having this new kind of failure state, at which point we can decide to handle it some other way (and stop occupying resources by repeatedly processing it).

(Incidentally, it seems like "aged out" could be another kind of failure status that should be formally noted.)

Reply:
I would rather punt on the "retry count exceeded" until another PR. This PR is already large enough. I've documented this in the README.